mbed_host_tests/host_tests/base_host_test.py

```diff
@@ -15,28 +15,171 @@
 limitations under the License.
 """
 
-class BaseHostTest():
+from time import time
+from inspect import isfunction, ismethod
+
+
+class BaseHostTestAbstract(object):
     """ Base class for each host-test test cases with standard
         setup, test and teardown set of functions
     """
 
     name = ''   # name of the host test (used for local registration)
+    __event_queue = None      # To main even loop
+    __dut_event_queue = None  # To DUT
 
-    def setup(self):
-        """ Set up function, initialize your test case dynamic resources in this function
+    def __notify_prn(self, text):
+        if self.__event_queue:
+            self.__event_queue.put(('__notify_prn', text, time()))
+
+    def __notify_conn_lost(self, text):
+        if self.__event_queue:
+            self.__event_queue.put(('__notify_conn_lost', text, time()))
+
+    def __notify_dut(self, key, value):
+        """! Send data over serial to DUT """
+        if self.__event_queue:
+            self.__dut_event_queue.put((key, value, time()))
+
+    def notify_complete(self, result=None):
+        """! Notify main even loop that host test finished processing
+        @param result True for success, False failure. If None - no action in main even loop
         """
+        if self.__event_queue:
+            self.__event_queue.put(('__notify_complete', result, time()))
+
+    def notify_conn_lost(self, text):
+        """! Notify main even loop that there was a DUT-host test connection error
+        @param consume If True htrun will process (consume) all remaining events
+        """
+        self.__notify_conn_lost(text)
+
+    def log(self, text):
+        """! Send log message to main event loop """
+        self.__notify_prn(text)
+
+    def send_kv(self, key, value):
+        """! Send Key-Value data to DUT """
+        self.__notify_dut(key, value)
+
+    def setup_communication(self, event_queue, dut_event_queue):
+        """! Setup queues used for IPC """
+        self.__event_queue = event_queue         # To main even loop
+        self.__dut_event_queue = dut_event_queue # To DUT
+
+    def setup(self):
+        """! Setup your tests and callbacks """
+        raise NotImplementedError
+
+    def result(self):
```

Comment:
Why do we have ```result()``` method when an async test can notify completion by calling ```notify_complete```. Two points to convey test results can cause buggy test implementation.